File: Roman Numeral Decoder.py
"""
Created on Fri Feb 17 09:17:13 2023

@author: Steven Cheng
"""

def function(roman):    

    ROMANS = {"I":1, "V":5, "X":10, "L":50, "C":100, "D":500, "M":1000}
    x = list()
    j=0

    if len(roman) == 1:
        x.append(ROMANS[roman[0]])

    elif len(roman) > 1:
        for i in range(len(roman)-1):
            if ROMANS[roman[i]] - ROMANS[roman[i+1]] < 0:
                x.append((ROMANS[roman[i+1]] - ROMANS[roman[i]]))
                j=1
            elif j==1:
                x.append(0)
                j=0
            else:
                x.append(ROMANS[roman[i]])

        if j==0:
            x.append(ROMANS[roman[len(roman)-1]])
    return(sum(x))

File: Roman Numeral Decoder.R
library("tidyverse")

solution <- function(roman){
  # complete the solution by transforming the roman numeral into an integer
  # steps,  1) convert string to a series of numbers.
  # 2) evaluate subtraction rule
  # 3) sum across numbers

  Symbol <- c("I", "V", "X", "L", "C", "D", "M")
  Value <- c(1,5,10,50,100,500,1000)
  
  Tbl <- as.data.frame(cbind(Symbol, Value))
  Tbl$Value <- as.numeric(Tbl$Value)
  
  ROMANS <- unlist(strsplit(roman,""))
  
  ROMANS <- Tbl$Value[match(ROMANS, Tbl$Symbol)]
  
  a <- c(ROMANS, NA)
  b <- c(NA, ROMANS)
  c <- ifelse(a-b > 0, a-b, NA)
  d <- c(c[-1], NA)
  
  sol <- ifelse(is.na(c)*is.na(d), a, c)
  
  return(sum(sol, na.rm = TRUE))
}